src/lib/util/printer.ml

```diff
@@ -0,0 +1,306 @@
+  | BG_Magenta -> "45"
+  | BG_Cyan -> "46"
+  | BG_Default -> "49"
+
+let style_of_tag t =
+  match Format_shims.get_stag t with
+  | "n" -> Normal
+  | "bold" -> Bold
+  | "/bold" -> Bold_off
+  | "underline" -> Underline
+  | "/underline" -> Underline_off
+
+  | "fg_black" -> FG_Black
+  | "fg_red" -> FG_Red
+  | "fg_green" -> FG_Green
+  | "fg_orange" -> FG_Orange
+  | "fg_blue" -> FG_Blue
+  | "fg_magenta" -> FG_Magenta
+  | "fg_cyan" -> FG_Cyan
+  | "fg_default" -> FG_Default
+
+  | "bg_black" -> BG_Black
+  | "bg_red" -> BG_Red
+  | "bg_green" -> BG_Green
+  | "bg_orange" -> BG_Orange
+  | "bg_blue" -> BG_Blue
+  | "bg_magenta" -> BG_Magenta
+  | "bg_cyan" -> BG_Cyan
+  | "bg_default" -> BG_Default
+
+  | _ -> raise Not_found
+
+let close_tag = function
+  | Bold -> Bold_off
+  | Underline -> Underline_off
+  | FG_Black | FG_Red | FG_Green | FG_Default -> FG_Default
+  | BG_Black | BG_Red | BG_Green | BG_Default -> BG_Default
+  | _ -> Normal
+
+let start_stag t =
+  try Printf.sprintf "\x1B[%sm" (to_value (style_of_tag t))
+  with Not_found -> ""
+
+let stop_stag t =
+  try Printf.sprintf "\x1B[%sm" (to_value (close_tag (style_of_tag t)))
+  with Not_found -> ""
+
+let add_colors formatter =
+  pp_set_tags formatter true;
+  let old_fs = Format_shims.pp_get_formatter_stag_functions formatter () in
+  Format_shims.pp_set_formatter_stag_functions formatter
+    (Format_shims.update_stag_functions old_fs start_stag stop_stag)
+
+let init_colors () =
+  add_colors (Options.get_fmt_std ());
+  add_colors (Options.get_fmt_wrn ());
+  add_colors (Options.get_fmt_err ());
+  add_colors (Options.get_fmt_dbg ())
+
+(************** Output Format *************)
+let clean_dbg_print = ref true
+let clean_wrn_print = ref true
+
+let pp_smt clean_print =
+  let smt = match Options.get_output_format () with
+    | Smtlib2 -> true
+    | Native | Why3 | Unknown _ -> false
+  in sprintf
+    (if smt && !clean_print then
+       begin clean_print := false; "@,; " end
+     else "")
+
+let pp_std_smt () =
+  match !clean_dbg_print, !clean_wrn_print with
+  | true, true -> ()
+  | false, true ->
+    clean_dbg_print := true;
+    fprintf (Options.get_fmt_std ()) "@,"
+  | true, false ->
+    clean_wrn_print := true;
+    fprintf (Options.get_fmt_std ()) "@,"
+  | false, false ->
+    clean_dbg_print := true;
+    clean_wrn_print := true;
+    fprintf (Options.get_fmt_std ()) "@,"
+
+let add_smt formatter =
+  let old_fs = Format_shims.pp_get_formatter_out_functions formatter () in
+  let out_newline () = old_fs.out_string "\n; " 0 3 in
+  Format_shims.pp_set_formatter_out_functions formatter
+    { old_fs with out_newline }
+
+let init_output_format () =
+  match Options.get_output_format () with
+  | Smtlib2 ->
+    add_smt (Options.get_fmt_wrn ());
+    add_smt (Options.get_fmt_dbg ())
+  | Native | Why3 | Unknown _ -> ()
+
+
+(************** Printers *************)
+let print_std s =
+  pp_std_smt ();
+  fprintf (Options.get_fmt_std ()) s
+
```

Comment:
This is where I would use the flush and why I tried to delete all the `@.` in the rest of the code. Since these functions have conditional printing, the easiest way is to use `kfprintf` who works with a callback. Let's first define a generic flush function:

```suggestion
let flush fmt = Format.fprintf fmt "@."

```